Finish refactoring IngestTypeVisitor, add datatype focused test

apmoriarty opened the pull request:
- completed refactoring datatype-gathering logic from the IngestTypePruningVisitor to the IngestTypeVisitor.
- add ShapesTest, a WiseGuys-style test with a focus on querying across datatypes
- updated IngestTypePruningVisitor and IngestTypeVisitor tests based on integration testing

foster33 commented on warehouse/query-core/src/test/java/datawave/query/ShapesTest.java:
```diff
@@ -0,0 +1,864 @@
+import datawave.webservice.query.configuration.GenericQueryConfiguration;
+
+/**
+ * A set of tests that emphasis the influence of datatypes on query planning and execution
```
```suggestion
 * A set of tests that emphasize the influence of datatypes on query planning and execution
```
Minor typo

apmoriarty committed:
Add ColorsTest for testing complex datatype, normalizer and tokenization interaction
Completed refactor of IngestTypeVisitor and IngestTypePruningVisitor
Changed how visitor handles negations and null terms

apmoriarty committed:
Clarify how ingest type results are used

apmoriarty committed:
add extra test case

alerman commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/IngestTypeVisitor.java:
```diff
@@ -82,11 +83,13 @@ public static Set<String> getIngestTypes(JexlNode node, TypeMetadata typeMetadat
-            if (!ingestTypes.contains(UNKNOWN_TYPE)) {
+            if (ingestTypes.contains(UNKNOWN_TYPE)) {
+                ingestTypes.retainAll(Collections.singleton(UNKNOWN_TYPE));
                 return ingestTypes;
```
Duplicated return. Remove this one and the flow doesnt change.

apmoriarty replied:
Not quite, in this case I only want to return a collection that contains the `UNKNOWN_TYPE`. I'll add a comment to that effect.

MiguelRicardos commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/IngestTypePruningVisitor.java:
```diff
@@ -324,7 +340,7 @@ private Set<String> pruneJunction(JexlNode node, Object data) {
             node.jjtGetChild(i).jjtAccept(this, data);
         }
-        return Collections.emptySet();
+        return new HashSet<>();
```
Is this actually more efficient?

Collections.emptySet() is an immutable object that isn't going to allocate more memory every time that this is called, whereas a new HashSet<>() is going to. Even if it's small, it seems like this could be bad for memory management.

apmoriarty replied:
The collection returned by `Collections.emptySet()` will throw an UnsupportedOperationException in some very unfortunate circumstances, so I ripped out all Collections and Guava utility methods in favor of "native" collections. Union and intersection operations are now done via `.addAll` and `.retain`

MiguelRicardos replied:
SGTM

alerman commented on warehouse/query-core/src/test/java/datawave/query/jexl/visitors/IngestTypePruningVisitorTest.java:
```diff
@@ -438,7 +441,7 @@ public void testEvaluationOnlyField() {
     public void testPruneNegation() {
         String query = "A == '1' || !((_Delayed_ = true) && (A == '1' && C == '2'))";
-        test(query, "A == '1'");
+        test(query, query);
```
Does this indicate we are not pruning the negation, if the expected now has the negation included?

apmoriarty replied:
Previously the negation would self-prune, let me take a closer look at why this stopped visiting the negated branch..

alerman reviewed (commented):
Just a few small things. Nothing too major. Please check the unit test with changed expected output.

MiguelRicardos commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/IngestTypePruningVisitor.java:
```diff
@@ -165,12 +156,33 @@ public Object visit(ASTReference node, Object data) {
+            node.jjtGetChild(i).jjtAccept(this, pruningTypes);
+        }
+
+        if (node.jjtGetNumChildren() == 0) {
```
Should this check be placed before the for loop?

apmoriarty replied:
No, a child of a union can self-prune in certain cases. Consider the query `(A && B) || (C && D)` where each term maps to a separate, unique datatype (no overlaps). The `(A && B)` branch will self prune to zero, and the `(C && D)` branch will prune to zero. Thus the union node now has zero children and must prune itself from it's parent.

MiguelRicardos replied:
Can you add a comment to explain that this will only happen in some instances? Do we have a unit test that shows this?

apmoriarty replied:
I'll add a comment. You can look at the `testDoubleNestedPruning()` unit test for some examples.

MiguelRicardos commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/IngestTypeVisitor.java:
```diff
@@ -354,25 +353,28 @@ public Set<String> getIngestTypesForField(String field) {
-                } else {
-                    ingestTypes = intersectTypes(ingestTypes, childIngestTypes);
-                }
+                ingestTypes = intersectTypes(ingestTypes, childIngestTypes);
```
Could you add a comment or explain why the special handling of negations for instances of ASTNNotNode's are no longer needed?

apmoriarty replied:
Does the method level comment answer the question about negations, or would you like to see something more explicit?

MiguelRicardos replied:
LGTM

MiguelRicardos commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/IngestTypeVisitor.java:
```diff
@@ -385,7 +387,9 @@ private Set<String> getIngestTypesForIntersection(ASTAndNode node) {
     private Set<String> intersectTypes(Set<String> typesA, Set<String> typesB) {
         if (typesA.contains(UNKNOWN_TYPE) || typesB.contains(UNKNOWN_TYPE)) {
-            return Collections.singleton(UNKNOWN_TYPE);
+            Set<String> unknown = new HashSet<>();
```
Could you explain the benefit of going from:
```
            return Collections.singleton(UNKNOWN_TYPE);
``` 
to 
```
Set<String> unknown = new HashSet<>();
unknown.add(UNKNOWN_TYPE);
return unknown;
```
Is there a reason to go from an immutable Set<String> to a mutable Set<String> if all it's doing is returning one with the UNKNOWN_TYPE?

apmoriarty replied:
The result will likely have set operations performed (add, addAll, retain) so it can't be immutable.

MiguelRicardos commented on warehouse/query-core/src/test/java/datawave/query/CompositeFunctionsTest.java:
```diff
@@ -491,7 +493,7 @@ public void testNotNulls() throws Exception {
                 "filter:isNotNull(ONE_NULL)",
                 "UUID =~ '^[CS].*' AND filter:isNotNull(UUID)",
-                "UUID =~ '^[CS].*' AND filter:isNotNull(NULL1)",
+                "UUID =~ '^[CS].*' AND filter:isNotNull(NULL1)",    //  dis guy
```
Could you add a more descriptive comment?

apmoriarty replied:
That's a leftover comment, I'll remove it

apmoriarty committed:
restore ability to perform an internal prune of a negated term

MiguelRicardos reviewed (approved):
Nice work!
